### PR TITLE
gossipsub / floodsub fixes

### PR DESCRIPTION
**Pull request by arnetheduck**
* remove timed cache - the window shifting already removes old messages
* ref -> object
* avoid unnecessary allocations with `[]` operator

### CODE REVIEW

**Comment by arnetheduck** [comment]
* floodsub, gossipsub: don't rebroadcast messages that fail validation
(!)
* floodsub, gossipsub: don't crash when unsubscribing from unknown
topics (!)
* gossipsub: don't send message to peers that are not interested in the
topic, when messages don't share topic list
* floodsub: don't repeat all messages for each message when
rebroadcasting
* floodsub: allow sending empty data
* floodsub: fix inefficient unsubscribe
* sync floodsub/gossipsub logging
* gossipsub: include incoming messages in mcache (!)
* gossipsub: don't rebroadcast already-seen messages (!)
* pubsubpeer: remove incoming/outgoing seen caches - these are already
handled in gossipsub, floodsub and will cause trouble when peers try to
resubscribe / regraft topics (because control messages will have same
digest)
* timedcache: reimplement without timers (fixes timer leaks and extreme
inefficiency due to per-message closures, futures etc)
* timedcache: ref -> obj

**Comment by sinkingsugar** [comment]
LGTM, do you have more gossip changes in the pipeline? I would like to know cos I wanna merge in 1.1

**Comment by arnetheduck** [comment]
@sinkingsugar #347 fixes some more things

File: libp2p/protocols/pubsub/floodsub.nim [hunk 1]
```diff
@@ -66,35 +65,34 @@ method rpcHandler*(f: FloodSub,
...
-      # forward the message to all peers interested in it
-      f.broadcast(
-        toSeq(toSendPeers),
-        RPCMsg(messages: rpcMsg.messages))
+      await handleData(f, t, msg.data)
 
-      trace "forwared message to peers", peers = toSendPeers.len
+    # In theory, if topics are the same in all messages, we could batch - we'd
+    # also have to be careful to only include validated messages
+    f.broadcast(toSeq(toSendPeers), RPCMsg(messages: @[msg]))
```
Comment (dryajov) [inline libp2p/protocols/pubsub/floodsub.nim:94]: Yes, we should potentially batch here, but should be OK for now.

**Review by dryajov — APPROVED** [review]
LGTM

**Pull request merged**